### PR TITLE
Separate browser tests into functional and content suites

### PR DESCRIPTION
**Pull request by chosak**
The current set of browser acceptance tests (run using `gulp test:acceptance`) include some that rely on a complete content database and some that can be run using a minimal set of generic content. This PR separates the tests into two different suites so that the functional subset can be easily run without needing a complete database dump.

## Additions

- Added suites to [test/browser_tests/environment.js](https://github.com/cfpb/cfgov-refresh/blob/cleanup-browser-tests/test/browser_tests/environment.js#L27) to group browser tests into logical groups.
- Added support for the `--suite` parameter when running `gulp test:acceptance`, for example:

 ```sh
 $ gulp test:acceptance --suite=functional
 $ gulp test:acceptance --suite=functional,content
 ```

 Support for `--specs` is still supported; however when passing `--suite`, the suite argument will take precedence. If neither of these arguments are provided, all tests will still run.

## Changes

- Updated `TEST.md` to describe new command-line usage.
- Modified `gulp test:acceptance` so that the `initial-test-data.sh` script is always run (removing a hardcoded check against the `wagtail` spec directory) to load data needed to pass the functional tests. This is suboptimal; it'd be more ideal to use a separate test database (maybe even the same Django DB as the unit tests) and should be revisited in future.

## Testing

- Run the functional browser test suite like this:

```sh
$ gulp test:acceptance --suite=functional
...
48 specs, 0 failures
Finished in 23.152 seconds
```

- The content-dependent browser test suite (which requires a complete database) can be run like this:

```sh
$ gulp test:acceptance --suite=content
...
255 specs, 0 failures, 27 pending specs
Finished in 100.738 seconds
```

- The full set of tests (all suites) can still be run like this:

```sh
$ gulp test:acceptance
...
303 specs, 0 failures, 27 pending specs
Finished in 121.615 seconds
```

- Individual or specific specs can still be run like this:

```sh
$ gulp test:acceptance --specs=foo*.js
...
5 specs, 0 failures
Finished in 2 seconds
```

- Running `gulp test` will now run the linter, the Python unit tests, and the functional test suite.

## Review

- @richa
- @sebworks

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

**Comment by chosak** [comment]
Trying to close and reopen this PR to see if it will [trigger a seemingly missed Travis build](http://stackoverflow.com/questions/19589260/force-travis-to-build-a-specific-pull-request-or-commit-sha-that-it-missed).

File: test/browser_tests/conf.js [hunk 1]
```diff
@@ -69,12 +69,31 @@ function _isSauceCredentialSet() {
...
+    var suitesArray = params.suite.split( ',' );
+    var specsArray = [];
+    for ( var i = 0, suite; suite = suitesArray[i++]; ) {
+      var suiteSpecs = environment.suites[suite];
+      if ( suiteSpecs.constructor === Array ) {
+        for ( var j = 0, len = suiteSpecs.length; j < len; j++ ) {
+          specsArray.push( suiteSpecs[j] );
+        }
+      } else {
+        specsArray.push(suiteSpecs);
```
Comment (anselmbradford) [inline test/browser_tests/conf.js]: Trivial: Add space in parens.

Comment (chosak) [inline test/browser_tests/conf.js]: Thanks, fixed.

**Comment by richaagarwal** [comment]
This is great, thanks Andy! Does `gulp test` still run everything?  Do we want to modify that?  Ideally we'd still have a single command people can reliably call.

**Comment by chosak** [comment]
@richaagarwal yes, `gulp test` still runs `lint`, `test:unit`, and `test:acceptance`. So right now the default is "run everything, including things that need the prod DB". I'd like to change Travis to add `gulp test:acceptance --suite:functional` so that it runs the non-content tests.

File: test/browser_tests/conf.js [hunk 2]
```diff
@@ -69,12 +69,31 @@ function _isSauceCredentialSet() {
...
   var parsedParams = {};
 
-  if ( _paramIsSet( params.specs ) ) {
+  // If one or more suites are specified, use their specs.
+  if ( _paramIsSet( params.suite ) ) {
+    var suitesArray = params.suite.split( ',' );
+    var specsArray = [];
+    for ( var i = 0, suite; suite = suitesArray[i++]; ) {
+      var suiteSpecs = environment.suites[suite];
+      if ( suiteSpecs.constructor === Array ) {
```
Comment (sebworks) [inline test/browser_tests/conf.js]: Probably safer to use `Array.isArray`. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray

File: gulp/tasks/test.js [hunk 3]
```diff
@@ -95,6 +95,9 @@ function _getProtractorParams() {
   // If --version=number flag is added on the command-line.
   params = _addCommandLineFlag( params, commandLineParams, 'version' );
 
+  // If --suite=suite1,suite2 flag is added on the command-line.
+  params = _addCommandLineFlag( params, commandLineParams, 'suite' );
+
   return params;
```
Comment (richaagarwal) [inline gulp/tasks/test.js]: Not a huge deal, but can we do something like `_addCommandLineFlags` (plural) here, that passes in 'version', 'suite', & all the other flags?

File: test/browser_tests/conf.js [hunk 4]
```diff
@@ -69,12 +69,31 @@ function _isSauceCredentialSet() {
...
+        for ( var j = 0, len = suiteSpecs.length; j < len; j++ ) {
+          specsArray.push( suiteSpecs[j] );
```
Comment (sebworks) [inline test/browser_tests/conf.js]: You could probably just use a hack to clone the array like `[1,2,3].slice(0)`. http://stackoverflow.com/questions/3978492/javascript-fastest-way-to-duplicate-an-array-slice-vs-for-loop

Comment (chosak) [inline test/browser_tests/conf.js]: Thanks @sebworks, `concat` is pretty useful, and much nicer than how I had been doing it.

File: gulp/tasks/test.js (hunk 3 above)
Comment (chosak) [inline gulp/tasks/test.js]: I agree that this code isn't the clearest; I left it as is just to keep the set of changes in this PR small.

**Comment by chosak** [comment]
Per @richaagarwal feedback, I changed this PR so that if you simply run `gulp test` from the command line, it'll now run only the functional browser tests (along with the Python unit tests and the linter). You can still run the full set of browser tests using `gulp test:acceptance` or with some of the command-line options described above.

**Comment by richaagarwal** [comment]
👍

File: gulp/tasks/test.js (hunk 3 above)
Comment (richaagarwal) [inline gulp/tasks/test.js]: Makes sense!

**Comment by sebworks** [comment]
Getting the following errors when running `gulp test:acceptance --suite=functional`:

screen shot 2016-07-22 at 10 47 59 am

**Comment by chosak** [comment]
@sebworks this might be an issue I encountered recently regarding a missing `HTTP_PORT` environment variable. Can you please try the solution described in the PR comments starting [here](https://github.com/cfpb/cfgov-refresh/pull/2225#issuecomment-232711269)?

**Comment by sebworks** [comment]
yep, that was the issue. Great work @chosak. 👍

**Pull request merged**